### PR TITLE
Add building stats

### PR DESCRIPTION
**Pull request by Raycoms**
Closes #
Closes #
Closes #

# Changes proposed in this pull request:
- Add building stats: 
- Forester by log type, farmer by crop type, planter by crop type, miner by block type.
- Library paper/book consumption and intelligence levels gained.
- Fisher just total fish.

[ ] Yes I tested this before submitting it.
[ ] I also did a multiplayer test.

Review please

### CODE REVIEW

File: src/main/java/com/minecolonies/coremod/entity/ai/basic/AbstractEntityAIInteract.java [hunk 1]
```diff
@@ -230,6 +231,15 @@ protected final boolean mineBlock(
         return true;
     }
 
+    /**
+     * Event triggered after receiving a list of block drops.
+     * @param blockDrops the received items from the block.
+     */
+    public void onBlockDropReception(final List<ItemStack> blockDrops)
+    {
+
```
Comment (Thodor12) [inline src/main/java/com/minecolonies/coremod/entity/ai/basic/AbstractEntityAIInteract.java]: Empty method? Should this have a // No-op text?

File: src/main/java/com/minecolonies/coremod/entity/ai/citizen/farmer/EntityAIWorkFarmer.java [hunk 2]
```diff
@@ -596,6 +596,16 @@ private boolean harvestIfAble(BlockPos position)
         return true;
     }
 
+    @Override
+    public void onBlockDropReception(final List<ItemStack> blockDrops)
+    {
+        super.onBlockDropReception(blockDrops);
+        for (final ItemStack stack : blockDrops)
+        {
+            building.getModule(STATS_MODULE).incrementBy( ITEM_OBTAINED + ";" + stack.getItem().getDescriptionId(), stack.getCount());
```
Comment (Thodor12) [inline src/main/java/com/minecolonies/coremod/entity/ai/citizen/farmer/EntityAIWorkFarmer.java]: There's some spacing here, formatting?

File: src/main/java/com/minecolonies/coremod/entity/ai/citizen/guard/KnightCombatAI.java [hunk 3]
```diff
@@ -352,6 +354,7 @@ protected void onTargetDied(final LivingEntity entity)
     {
         parentAI.incrementActionsDoneAndDecSaturation();
         user.getCitizenExperienceHandler().addExperience(EXP_PER_MOB_DEATH);
-        user.getCitizenColonyHandler().getColony().getStatisticsManager().increment(MOBS_KILLED);
+        user.getCitizenColonyHandler().getColony().getStatisticsManager().increment(MOBS_KILLED, user.getCitizenColonyHandler().getColony().getDay());
+        parentAI.building.getModule(STATS_MODULE).increment( MOB_KILLED + ";" + entity.getType().getDescription().toString());
```
Comment (Thodor12) [inline src/main/java/com/minecolonies/coremod/entity/ai/citizen/guard/KnightCombatAI.java]: Spacing again

File: src/main/java/com/minecolonies/coremod/entity/ai/citizen/guard/RangerCombatAI.java [hunk 4]
```diff
@@ -359,6 +361,7 @@ protected void onTargetDied(final LivingEntity entity)
     {
         parentAI.incrementActionsDoneAndDecSaturation();
         user.getCitizenExperienceHandler().addExperience(EXP_PER_MOB_DEATH);
-        user.getCitizenColonyHandler().getColony().getStatisticsManager().increment(MOBS_KILLED);
+        user.getCitizenColonyHandler().getColony().getStatisticsManager().increment(MOBS_KILLED, user.getCitizenColonyHandler().getColony().getDay());
+        parentAI.building.getModule(STATS_MODULE).increment( MOB_KILLED + ";" + entity.getType().getDescription().toString());
```
Comment (Thodor12) [inline src/main/java/com/minecolonies/coremod/entity/ai/citizen/guard/RangerCombatAI.java]: Spacing again

File: src/main/java/com/minecolonies/coremod/entity/ai/citizen/lumberjack/EntityAIWorkLumberjack.java [hunk 5]
```diff
@@ -568,6 +569,16 @@ else if (job.getTree().hasLeaves() && shouldBreakLeaves)
         return getState();
     }
 
+    @Override
+    public void onBlockDropReception(final List<ItemStack> blockDrops)
+    {
+        super.onBlockDropReception(blockDrops);
+        for (final ItemStack stack : blockDrops)
+        {
+            building.getModule(STATS_MODULE).incrementBy( ITEM_OBTAINED + ";" + stack.getItem().getDescriptionId(), stack.getCount());
```
Comment (Thodor12) [inline src/main/java/com/minecolonies/coremod/entity/ai/citizen/lumberjack/EntityAIWorkLumberjack.java]: Spacing again

File: src/main/java/com/minecolonies/coremod/entity/ai/citizen/miner/EntityAIQuarrier.java [hunk 6]
```diff
@@ -201,6 +202,16 @@ private Tuple<String, String> getShaftPath(@NotNull final IBuilding quarry)
         return new Tuple<>(path, name);
     }
 
+    @Override
+    public void onBlockDropReception(final List<ItemStack> blockDrops)
+    {
+        super.onBlockDropReception(blockDrops);
+        for (final ItemStack stack : blockDrops)
+        {
+            building.getModule(STATS_MODULE).incrementBy( ITEM_OBTAINED + ";" + stack.getItem().getDescriptionId(), stack.getCount());
```
Comment (Thodor12) [inline src/main/java/com/minecolonies/coremod/entity/ai/citizen/miner/EntityAIQuarrier.java]: Spacing again

File: src/main/java/com/minecolonies/coremod/entity/ai/citizen/miner/EntityAIStructureMiner.java [hunk 7]
```diff
@@ -891,6 +891,16 @@ public void executeSpecificCompleteActions()
         job.setBlueprint(null);
     }
 
+    @Override
+    public void onBlockDropReception(final List<ItemStack> blockDrops)
+    {
+        super.onBlockDropReception(blockDrops);
+        for (final ItemStack stack : blockDrops)
+        {
+            building.getModule(STATS_MODULE).incrementBy( ITEM_OBTAINED + ";" + stack.getItem().getDescriptionId(), stack.getCount());
```
Comment (Thodor12) [inline src/main/java/com/minecolonies/coremod/entity/ai/citizen/miner/EntityAIStructureMiner.java]: Spacing again

File: src/main/java/com/minecolonies/coremod/entity/ai/citizen/planter/EntityAIWorkPlanter.java [hunk 8]
```diff
@@ -140,6 +142,16 @@ private IAIState pickField()
         return IDLE;
     }
 
+    @Override
+    public void onBlockDropReception(final List<ItemStack> blockDrops)
+    {
+        super.onBlockDropReception(blockDrops);
+        for (final ItemStack stack : blockDrops)
+        {
+            building.getModule(STATS_MODULE).incrementBy( ITEM_OBTAINED + ";" + stack.getItem().getDescriptionId(), stack.getCount());
```
Comment (Thodor12) [inline src/main/java/com/minecolonies/coremod/entity/ai/citizen/planter/EntityAIWorkPlanter.java]: Spacing again

File: src/main/java/com/minecolonies/coremod/entity/ai/citizen/student/EntityAIStudy.java [hunk 9]
```diff
@@ -152,11 +155,15 @@ private IAIState study()
...
-            data.getCitizenSkillHandler().tryLevelUpIntelligence(data.getRandom(), ONE_IN_X_CHANCE * (100D / chosenItem.getSkillIncreasePct()), data);
+            if (data.getCitizenSkillHandler().tryLevelUpIntelligence(data.getRandom(), ONE_IN_X_CHANCE * (100D / chosenItem.getSkillIncreasePct()), data))
+            {
+                building.getModule(STATS_MODULE).increment(INT_LEVELED);
+            }
             // Break item rand
             if (world.random.nextInt(100) <= chosenItem.getBreakPct())
             {
                 data.getInventory().extractItem(chosenItem.getSlot(), 1, false);
+                building.getModule(STATS_MODULE).increment( ITEM_USED + ";" + chosenItem.getItem().getDescriptionId());
```
Comment (Thodor12) [inline src/main/java/com/minecolonies/coremod/entity/ai/citizen/student/EntityAIStudy.java]: Spacing again

**Review by Thodor12 — DISMISSED** [review]
Looks good, some minor nitpicks

File: src/main/java/com/minecolonies/coremod/entity/ai/basic/AbstractEntityAIInteract.java (hunk 1 above)
Comment (Raycoms) [reply src/main/java/com/minecolonies/coremod/entity/ai/basic/AbstractEntityAIInteract.java]: Default method, other override if needed.

Comment (Thodor12) [reply src/main/java/com/minecolonies/coremod/entity/ai/basic/AbstractEntityAIInteract.java]: Still, maybe just a simple `// No-op` text in it like we do in the majority of places just as an indication this is purposefully left empty.

**Review by Thodor12 — APPROVED** [review]

**Review by someaddons — APPROVED** [review]

**Pull request merged**